File: android/app/src/main/java/com/noop/analytics/TrainingLoadModel.kt
package com.noop.analytics

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Transparent calendar-day training-load context; the Kotlin twin of `TrainingLoadModel.swift`.
 *
 * Every input must use the same additive unit (for example session-RPE minutes, TRIMP, or
 * MET-minutes). A bounded WHOOP/NOOP strain score is nonlinear and must not be summed as an impulse
 * load. ATL/CTL are descriptive exponentially weighted averages and TSB is exactly `CTL - ATL`;
 * none directly measures fitness, fatigue, readiness, injury risk, or permission to train.
 *
 * Formula and cold-start choices were independently implemented and cross-checked against the MIT
 * OpenStrap analytics implementation at commit `ebe45da6271ec99fbda353067b6deadc757ff2d8`
 * and the MIT `RuochenLyu/apple-health-analyst` project. No source code was copied.
 */
object TrainingLoadModel {
    /** Shared hard ceiling for every day-count/time-horizon configuration and timeline allocation. */
    private const val MAX_CONFIGURED_DAYS = 36_600

    data class Entry(val day: String, val load: Double)
    data class DailyLoad(val day: String, val load: Double)
    data class Aggregation(val days: List<DailyLoad>, val rejectedEntryCount: Int)

    enum class MissingDayPolicy {
        /** Abstain on gaps and restart warm-up; never manufacture zero load. */
        REQUIRE_OBSERVED,

        /** Explicitly use zero for gaps, while marking affected output as estimated. */
        ASSUME_REST,
    }

    data class Configuration(
        val acuteTimeConstantDays: Double = 7.0,
        val chronicTimeConstantDays: Double = 42.0,
        val minimumHistoryDays: Int = 14,
        val seedDays: Int = 7,
        val rampWindowDays: Int = 7,
        val stableRampFraction: Double = 0.10,
        /** Expanded daily timeline bound; 10 years by default, with a hard 100-year ceiling. */
        val maximumCalendarSpanDays: Int = 3_660,
        val missingDayPolicy: MissingDayPolicy = MissingDayPolicy.REQUIRE_OBSERVED,
    ) {
        init {
            require(
                acuteTimeConstantDays > 0 && acuteTimeConstantDays.isFinite() &&
                    acuteTimeConstantDays <= MAX_CONFIGURED_DAYS.toDouble()
            )
            require(
                chronicTimeConstantDays > 0 && chronicTimeConstantDays.isFinite() &&
                    chronicTimeConstantDays <= MAX_CONFIGURED_DAYS.toDouble()
            )
            require(minimumHistoryDays in 1..MAX_CONFIGURED_DAYS)
            require(seedDays in 1..minimumHistoryDays)
            require(rampWindowDays in 1..MAX_CONFIGURED_DAYS)
            require(stableRampFraction >= 0 && stableRampFraction.isFinite())
            require(maximumCalendarSpanDays in 1..MAX_CONFIGURED_DAYS)
        }
    }

    enum class DaySource { OBSERVED, MISSING, ASSUMED_REST }
    enum class Quality { UNAVAILABLE, PROVISIONAL, OBSERVED, ESTIMATED }
    enum class RampDirection { BUILDING, STEADY, EASING, UNAVAILABLE }

    data class Point(
        val day: String,
        val observedLoad: Double?,
        val effectiveLoad: Double?,
        val source: DaySource,
        val atl: Double?,
        val ctl: Double?,
        val tsb: Double?,
        val currentRampWindowLoad: Double?,
        val previousRampWindowLoad: Double?,
        val rampChange: Double?,
        val rampChangeFraction: Double?,
        val rampDirection: RampDirection,
        val observedCoverage: Double,
        val assumedRestDaysInWindow: Int,
        /** Assumptions still folded into the recursive ATL/CTL model state. */
        val assumedRestDaysInModelHistory: Int,
        val consecutiveHistoryDays: Int,
        val quality: Quality,
        val interpretation: String,
    )

    data class Series(
        val observedDays: List<DailyLoad>,
        val points: List<Point>,
        val rejectedEntryCount: Int,
        val status: Status,
    ) {
        enum class Status { COMPLETE, EMPTY, CALENDAR_SPAN_EXCEEDED }
        val latest: Point? get() = points.lastOrNull()
    }

    /** Validate inputs and add all valid load contributions sharing one day. */
    fun aggregate(entries: List<Entry>): Aggregation {
        val totals = mutableMapOf<String, Double>()
        var rejected = 0
        entries.forEach { entry ->
            if (parseDay(entry.day) == null || !entry.load.isFinite() || entry.load < 0) {
                rejected += 1
            } else {
                val total = (totals[entry.day] ?: 0.0) + entry.load
                if (!total.isFinite()) {
                    // Keep the valid total so far and reject only the overflowing contribution.
                    rejected += 1
                } else {
                    totals[entry.day] = total
                }
            }
        }
        val days = totals.keys.sorted().map { day -> DailyLoad(day, totals.getValue(day)) }
        return Aggregation(days, rejected)
    }

    fun evaluate(
        entries: List<Entry>,
        configuration: Configuration = Configuration(),
    ): Series {
        val aggregation = aggregate(entries)
        val empty = Series(aggregation.days, emptyList(), aggregation.rejectedEntryCount, Series.Status.EMPTY)
        val first = aggregation.days.firstOrNull()?.day?.let(::parseDay) ?: return empty
        val last = aggregation.days.lastOrNull()?.day?.let(::parseDay) ?: return empty

        val span = ChronoUnit.DAYS.between(first, last) + 1
        if (span <= 0 || span > configuration.maximumCalendarSpanDays) {
            return Series(
                aggregation.days,
                emptyList(),
                aggregation.rejectedEntryCount,
                Series.Status.CALENDAR_SPAN_EXCEEDED,
            )
        }

        val observedByDay = aggregation.days.associate { it.day to it.load }
        val timeline = buildList {
            var date = first
            while (!date.isAfter(last)) {
                val key = date.toString()
                add(key to observedByDay[key])
                date = date.plusDays(1)
            }
        }

        val acuteAlpha = 1 - exp(-1 / configuration.acuteTimeConstantDays)
        val chronicAlpha = 1 - exp(-1 / configuration.chronicTimeConstantDays)
        val coverageWindow = max(1, configuration.chronicTimeConstantDays.roundToInt())
        val segmentLoads = mutableListOf<Double>()
        var atlState: Double? = null
        var ctlState: Double? = null
        var segmentAssumedDays = 0
        val effectiveLoads = mutableListOf<Double?>()
        val sources = mutableListOf<DaySource>()
        val points = mutableListOf<Point>()

        timeline.forEachIndexed { index, item ->
            val observed = item.second
            val source: DaySource
            val effective: Double?
            if (observed != null) {
                source = DaySource.OBSERVED
                effective = observed
            } else if (configuration.missingDayPolicy == MissingDayPolicy.ASSUME_REST) {
                source = DaySource.ASSUMED_REST
                effective = 0.0
            } else {
                source = DaySource.MISSING
                effective = null
            }
            effectiveLoads += effective
            sources += source

            if (effective != null) {
                segmentLoads += effective
                if (source == DaySource.ASSUMED_REST) segmentAssumedDays += 1
                if (segmentLoads.size == configuration.seedDays) {
                    val seed = finiteMean(segmentLoads)
                    if (seed != null) {
                        atlState = seed
                        ctlState = seed
                    } else {
                        // Never expose a non-finite recursive state as a health/load estimate.
                        segmentLoads.clear()
                        segmentAssumedDays = 0
                        atlState = null
                        ctlState = null
                    }
                } else if (segmentLoads.size > configuration.seedDays) {
                    val nextAtl = atlState?.let { it + acuteAlpha * (effective - it) }
                    val nextCtl = ctlState?.let { it + chronicAlpha * (effective - it) }
                    if (nextAtl != null && nextCtl != null && nextAtl.isFinite() && nextCtl.isFinite()) {
                        atlState = nextAtl
                        ctlState = nextCtl
                    } else {
                        segmentLoads.clear()
                        segmentAssumedDays = 0
                        atlState = null
                        ctlState = null
                    }
                }
            } else {
                // Unknown load makes the exact EWMA unknowable. Strict mode abstains and starts a new
                // warm-up segment rather than silently applying zero or compressing calendar time.
                segmentLoads.clear()
                segmentAssumedDays = 0
                atlState = null
                ctlState = null
            }

            val modelIsWarm = segmentLoads.size >= configuration.minimumHistoryDays
            val candidateAtl = if (modelIsWarm) atlState?.takeIf { it.isFinite() } else null
            val candidateCtl = if (modelIsWarm) ctlState?.takeIf { it.isFinite() } else null
            val candidateTsb = if (candidateCtl != null && candidateAtl != null) {
                (candidateCtl - candidateAtl).takeIf { it.isFinite() }
            } else {
                null
            }
            // ATL/CTL/TSB are one coherent estimate. If any derived member is non-finite, publish none.
            val atl = if (candidateTsb != null) candidateAtl else null
            val ctl = if (candidateTsb != null) candidateCtl else null
            val tsb = candidateTsb
            val ramp = rampContext(
                effectiveLoads,
                configuration.rampWindowDays,
                configuration.stableRampFraction,
            )
            val coverageStart = max(0, index - coverageWindow + 1)
            val coverageSources = sources.subList(coverageStart, index + 1)
            val observedCount = coverageSources.count { it == DaySource.OBSERVED }
            val assumedCount = coverageSources.count { it == DaySource.ASSUMED_REST }
            val coverage = observedCount.toDouble() / coverageSources.size
            val quality = when {
                atl == null || ctl == null -> Quality.UNAVAILABLE
                segmentAssumedDays > 0 -> Quality.ESTIMATED
                segmentLoads.size < coverageWindow -> Quality.PROVISIONAL
                else -> Quality.OBSERVED
            }

            points += Point(
                day = item.first,
                observedLoad = observed,
                effectiveLoad = effective,
                source = source,
                atl = atl,
                ctl = ctl,
                tsb = tsb,
                currentRampWindowLoad = ramp.current,
                previousRampWindowLoad = ramp.previous,
                rampChange = ramp.change,
                rampChangeFraction = ramp.fraction,
                rampDirection = ramp.direction,
                observedCoverage = coverage,
                assumedRestDaysInWindow = assumedCount,
                assumedRestDaysInModelHistory = segmentAssumedDays,
                consecutiveHistoryDays = segmentLoads.size,
                quality = quality,
                interpretation = interpretation(
                    quality, tsb, segmentLoads.size, configuration.minimumHistoryDays, segmentAssumedDays,
                ),
            )
        }

        return Series(aggregation.days, points, aggregation.rejectedEntryCount, Series.Status.COMPLETE)
    }

    private data class RampContext(
        val current: Double? = null,
        val previous: Double? = null,
        val change: Double? = null,
        val fraction: Double? = null,
        val direction: RampDirection = RampDirection.UNAVAILABLE,
    )

    private fun rampContext(
        effectiveLoads: List<Double?>,
        windowDays: Int,
        stableFraction: Double,
    ): RampContext {
        // Division-first avoids `windowDays * 2` overflow even if this helper is reused independently
        // of Configuration's public hard bounds.
        if (windowDays <= 0 || windowDays > effectiveLoads.size / 2) return RampContext()
        val currentStart = effectiveLoads.size - windowDays
        val previousStart = currentStart - windowDays
        val previousSlice = effectiveLoads.subList(previousStart, currentStart)
        val currentSlice = effectiveLoads.subList(currentStart, effectiveLoads.size)
        if (previousSlice.any { it == null } || currentSlice.any { it == null }) return RampContext()
        val previous = finiteSum(previousSlice.filterNotNull()) ?: return RampContext()
        val current = finiteSum(currentSlice.filterNotNull()) ?: return RampContext()
        val change = current - previous
        if (!change.isFinite()) return RampContext()
        if (previous <= 0) {
            return RampContext(
                current, previous, change, null,
                if (current > 0) RampDirection.BUILDING else RampDirection.STEADY,
            )
        }
        val fraction = change / previous
        if (!fraction.isFinite()) {
            return RampContext(current, previous, change, null, RampDirection.UNAVAILABLE)
        }
        val direction = when {
            fraction > stableFraction -> RampDirection.BUILDING
            fraction < -stableFraction -> RampDirection.EASING
            else -> RampDirection.STEADY
        }
        return RampContext(current, previous, change, fraction, direction)
    }

    private fun interpretation(
        quality: Quality,
        tsb: Double?,
        consecutiveDays: Int,
        minimumDays: Int,
        assumedCount: Int,
    ): String {
        if (tsb == null || !tsb.isFinite()) {
            return "Load comparison unavailable: $consecutiveDays of $minimumDays consecutive calendar days recorded."
        }
        if (quality == Quality.ESTIMATED) {
            val suffix = if (assumedCount == 1) "" else "s"
            return "Estimate includes $assumedCount unobserved day$suffix treated as rest; it describes recorded load only."
        }
        val comparison = when {
            tsb < -0.000_001 -> "Recent recorded load is above the longer-term recorded-load baseline."
            tsb > 0.000_001 -> "Recent recorded load is below the longer-term recorded-load baseline."
            else -> "Recent and longer-term recorded load are similar."
        }
        return if (quality == Quality.PROVISIONAL) {
            "Provisional: $comparison This is load context, not a fitness or fatigue measurement."
        } else {
            "$comparison This is load context, not a fitness or fatigue measurement."
        }
    }

    /** Overflow-safe mean for nonnegative finite daily loads. */
    private fun finiteMean(values: List<Double>): Double? {
        if (values.isEmpty()) return null
        var mean = 0.0
        values.forEachIndexed { index, value ->
            if (!value.isFinite() || value < 0) return null
            val next = mean + (value - mean) / (index + 1).toDouble()
            if (!next.isFinite()) return null
            mean = next
        }
        return mean
    }

    /** Return null rather than publishing Infinity when a ramp-window total is unrepresentable. */
    private fun finiteSum(values: List<Double>): Double? {
        var total = 0.0
        for (value in values) {
            if (!value.isFinite() || value < 0) return null
            val next = total + value
            if (!next.isFinite()) return null
            total = next
        }
        return total
    }

    private fun parseDay(value: String): LocalDate? =
        runCatching { LocalDate.parse(value).takeIf { it.toString() == value } }.getOrNull()
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/TrainingLoadModel.swift
import Foundation

/// Transparent calendar-day training-load context.
///
/// Inputs must use one *additive* load unit throughout a series (for example session-RPE minutes,
/// TRIMP, or MET-minutes). Multiple sessions on the same day are summed. A bounded headline score
/// such as WHOOP/NOOP strain is nonlinear and must not be summed as though it were an impulse load.
///
/// ATL and CTL are exponentially weighted averages with 7- and 42-day time constants by default:
/// `alpha = 1 - exp(-1 / timeConstant)`. TSB is the descriptive difference `CTL - ATL`. These values
/// compare recent *recorded load* with a longer recorded-load baseline; they do not directly measure
/// fatigue, fitness, readiness, injury risk, or permission to train.
///
/// Formula and cold-start choices were independently implemented and cross-checked against the MIT
/// OpenStrap analytics implementation at commit `ebe45da6271ec99fbda353067b6deadc757ff2d8`
/// and the MIT `RuochenLyu/apple-health-analyst` project. No source code was copied.
public enum TrainingLoadModel {

    /// All day-count configuration stays within the same hard ceiling as timeline expansion. Besides
    /// bounding memory, this keeps every count conversion and window-index calculation far below `Int`
    /// overflow on every supported architecture.
    private static let maximumConfiguredDays = 36_600

    /// One additive load contribution. Several entries may share a day.
    public struct Entry: Sendable, Equatable {
        public let day: String
        public let load: Double

        public init(day: String, load: Double) {
            self.day = day
            self.load = load
        }
    }

    /// An observed calendar-day total. `0` is an explicitly observed rest day.
    public struct DailyLoad: Sendable, Equatable {
        public let day: String
        public let load: Double

        public init(day: String, load: Double) {
            self.day = day
            self.load = load
        }
    }

    public struct Aggregation: Sendable, Equatable {
        public let days: [DailyLoad]
        /// Entries rejected because their day was not canonical `YYYY-MM-DD`, or load was negative/non-finite.
        public let rejectedEntryCount: Int
    }

    /// Absence is unknown by default. Calling code must opt in before interpreting absent days as rest.
    public enum MissingDayPolicy: Sendable, Equatable {
        /// Do not manufacture zero load. A gap makes that day unavailable and restarts EWMA warm-up.
        case requireObserved
        /// Use zero for absent calendar days, while marking every affected output as estimated.
        case assumeRest
    }

    public struct Configuration: Sendable, Equatable {
        public let acuteTimeConstantDays: Double
        public let chronicTimeConstantDays: Double
        public let minimumHistoryDays: Int
        public let seedDays: Int
        public let rampWindowDays: Int
        public let stableRampFraction: Double
        /// Maximum inclusive first-to-last calendar span expanded into daily points. Defaults to ten
        /// years and is hard-capped at 100 years so untrusted dates cannot trigger unbounded allocation.
        public let maximumCalendarSpanDays: Int
        public let missingDayPolicy: MissingDayPolicy

        public init(acuteTimeConstantDays: Double = 7,
                    chronicTimeConstantDays: Double = 42,
                    minimumHistoryDays: Int = 14,
                    seedDays: Int = 7,
                    rampWindowDays: Int = 7,
                    stableRampFraction: Double = 0.10,
                    maximumCalendarSpanDays: Int = 3_660,
                    missingDayPolicy: MissingDayPolicy = .requireObserved) {
            precondition(acuteTimeConstantDays > 0 && acuteTimeConstantDays.isFinite
                         && acuteTimeConstantDays <= Double(TrainingLoadModel.maximumConfiguredDays))
            precondition(chronicTimeConstantDays > 0 && chronicTimeConstantDays.isFinite
                         && chronicTimeConstantDays <= Double(TrainingLoadModel.maximumConfiguredDays))
            precondition((1...TrainingLoadModel.maximumConfiguredDays).contains(minimumHistoryDays))
            precondition(seedDays >= 1 && seedDays <= minimumHistoryDays)
            precondition((1...TrainingLoadModel.maximumConfiguredDays).contains(rampWindowDays))
            precondition(stableRampFraction >= 0 && stableRampFraction.isFinite)
            precondition((1...TrainingLoadModel.maximumConfiguredDays).contains(maximumCalendarSpanDays))
            self.acuteTimeConstantDays = acuteTimeConstantDays
            self.chronicTimeConstantDays = chronicTimeConstantDays
            self.minimumHistoryDays = minimumHistoryDays
            self.seedDays = seedDays
            self.rampWindowDays = rampWindowDays
            self.stableRampFraction = stableRampFraction
            self.maximumCalendarSpanDays = maximumCalendarSpanDays
            self.missingDayPolicy = missingDayPolicy
        }
    }

    public enum DaySource: String, Sendable, Equatable {
        case observed
        case missing
        case assumedRest
    }

    public enum Quality: String, Sendable, Equatable {
        case unavailable
        case provisional
        case observed
        case estimated
    }

    public enum RampDirection: String, Sendable, Equatable {
        case building
        case steady
        case easing
        case unavailable
    }

    public struct Point: Sendable, Equatable {
        public let day: String
        /// The recorded daily total. Nil means no record existed for this calendar day.
        public let observedLoad: Double?
        /// The value folded into the model. Nil in strict mode; zero for an assumed rest day.
        public let effectiveLoad: Double?
        public let source: DaySource
        /// 7-day exponentially weighted recent load (default horizon); nil during warm-up/gaps.
        public let atl: Double?
        /// 42-day exponentially weighted load baseline (default horizon); nil during warm-up/gaps.
        public let ctl: Double?
        /// `CTL - ATL`, using the same day's estimates.
        public let tsb: Double?
        public let currentRampWindowLoad: Double?
        public let previousRampWindowLoad: Double?
        public let rampChange: Double?
        /// `(current - previous) / previous`; nil when the previous window total is zero.
        public let rampChangeFraction: Double?
        public let rampDirection: RampDirection
        /// Fraction of actually observed days in the trailing chronic-horizon calendar window.
        public let observedCoverage: Double
        public let assumedRestDaysInWindow: Int
        /// Assumptions folded into the current EWMA segment. Unlike window coverage, this stays nonzero
        /// while an assumed impulse can still influence the recursive state.
        public let assumedRestDaysInModelHistory: Int
        /// Consecutive effective calendar days behind this estimate; resets after a strict missing day.
        public let consecutiveHistoryDays: Int
        public let quality: Quality
        /// Cautious UI-ready context. It deliberately avoids fitness/fatigue/injury claims.
        public let interpretation: String
    }

    public struct Series: Sendable, Equatable {
        public enum Status: String, Sendable, Equatable {
            case complete
            case empty
            case calendarSpanExceeded
        }

        /// Observed daily totals only; missing calendar dates are represented in `points`.
        public let observedDays: [DailyLoad]
        /// One point for every calendar day from first to last valid input.
        public let points: [Point]
        public let rejectedEntryCount: Int
        public let status: Status

        public var latest: Point? { points.last }
    }

    /// Validate entries and add all valid contributions that share a calendar day.
    public static func aggregate(_ entries: [Entry]) -> Aggregation {
        var totals: [String: Double] = [:]
        var rejected = 0
        for entry in entries {
            guard parseDay(entry.day) != nil, entry.load.isFinite, entry.load >= 0 else {
                rejected += 1
                continue
            }
            let total = totals[entry.day, default: 0] + entry.load
            guard total.isFinite else {
                // Keep the valid total accumulated so far and reject only the contribution that overflowed.
                rejected += 1
                continue
            }
            totals[entry.day] = total
        }
        let days = totals.keys.sorted().compactMap { day -> DailyLoad? in
            totals[day].map { DailyLoad(day: day, load: $0) }
        }
        return Aggregation(days: days, rejectedEntryCount: rejected)
    }

    public static func evaluate(entries: [Entry],
                                configuration: Configuration = Configuration()) -> Series {
        let aggregation = aggregate(entries)
        guard let first = aggregation.days.first.flatMap({ parseDay($0.day) }),
              let last = aggregation.days.last.flatMap({ parseDay($0.day) }) else {
            return Series(observedDays: aggregation.days, points: [],
                          rejectedEntryCount: aggregation.rejectedEntryCount, status: .empty)
        }

        let observedByDay = Dictionary(uniqueKeysWithValues: aggregation.days.map { ($0.day, $0.load) })
        let calendar = utcCalendar
        let span = calendar.dateComponents([.day], from: first, to: last).day.map { $0 + 1 }
        guard let span, span > 0, span <= configuration.maximumCalendarSpanDays else {
            return Series(observedDays: aggregation.days, points: [],
                          rejectedEntryCount: aggregation.rejectedEntryCount,
                          status: .calendarSpanExceeded)
        }
        var timeline: [(day: String, observed: Double?)] = []
        timeline.reserveCapacity(span)
        var date = first
        while date <= last {
            let key = dayKey(date)
            timeline.append((key, observedByDay[key]))
            guard let next = calendar.date(byAdding: .day, value: 1, to: date) else { break }
            date = next
        }

        let acuteAlpha = 1 - exp(-1 / configuration.acuteTimeConstantDays)
        let chronicAlpha = 1 - exp(-1 / configuration.chronicTimeConstantDays)
        let coverageWindow = max(1, Int(configuration.chronicTimeConstantDays.rounded()))
        var segmentLoads: [Double] = []
        var atlState: Double?
        var ctlState: Double?
        var segmentAssumedDays = 0
        var effectiveLoads: [Double?] = []
        var sources: [DaySource] = []
        var points: [Point] = []

        for (index, item) in timeline.enumerated() {
            let source: DaySource
            let effective: Double?
            if let observed = item.observed {
                source = .observed
                effective = observed
            } else if configuration.missingDayPolicy == .assumeRest {
                source = .assumedRest
                effective = 0
            } else {
                source = .missing
                effective = nil
            }
            effectiveLoads.append(effective)
            sources.append(source)

            if let load = effective {
                segmentLoads.append(load)
                if source == .assumedRest { segmentAssumedDays += 1 }
                if segmentLoads.count == configuration.seedDays {
                    if let seed = finiteMean(segmentLoads) {
                        atlState = seed
                        ctlState = seed
                    } else {
                        // A finite input can still make an intermediate overflow if a future formula
                        // changes. Abstain and restart rather than publishing NaN/Infinity as health data.
                        segmentLoads.removeAll(keepingCapacity: true)
                        segmentAssumedDays = 0
                        atlState = nil
                        ctlState = nil
                    }
                } else if segmentLoads.count > configuration.seedDays,
                          let previousAtl = atlState, let previousCtl = ctlState {
                    let nextAtl = previousAtl + acuteAlpha * (load - previousAtl)
                    let nextCtl = previousCtl + chronicAlpha * (load - previousCtl)
                    if nextAtl.isFinite, nextCtl.isFinite {
                        atlState = nextAtl
                        ctlState = nextCtl
                    } else {
                        segmentLoads.removeAll(keepingCapacity: true)
                        segmentAssumedDays = 0
                        atlState = nil
                        ctlState = nil
                    }
                }
            } else {
                // A mathematically exact EWMA cannot be recovered when its input is unknown. Strict mode
                // therefore abstains and starts a new warm-up segment after the gap instead of silently
                // applying a zero impulse or compressing calendar time.
                segmentLoads.removeAll(keepingCapacity: true)
                segmentAssumedDays = 0
                atlState = nil
                ctlState = nil
            }

            let modelIsWarm = segmentLoads.count >= configuration.minimumHistoryDays
            let atl: Double?
            let ctl: Double?
            let tsb: Double?
            if modelIsWarm, let candidateAtl = atlState, let candidateCtl = ctlState,
               candidateAtl.isFinite, candidateCtl.isFinite {
                let candidateTsb = candidateCtl - candidateAtl
                if candidateTsb.isFinite {
                    atl = candidateAtl
                    ctl = candidateCtl
                    tsb = candidateTsb
                } else {
                    atl = nil
                    ctl = nil
                    tsb = nil
                }
            } else {
                atl = nil
                ctl = nil
                tsb = nil
            }

            let ramp = rampContext(effectiveLoads: effectiveLoads,
                                   windowDays: configuration.rampWindowDays,
                                   stableFraction: configuration.stableRampFraction)
            let coverageStart = max(0, index - coverageWindow + 1)
            let coverageSources = sources[coverageStart...index]
            let observedCount = coverageSources.filter { $0 == .observed }.count
            let assumedCount = coverageSources.filter { $0 == .assumedRest }.count
            let coverage = Double(observedCount) / Double(coverageSources.count)
            let quality: Quality
            if atl == nil || ctl == nil {
                quality = .unavailable
            } else if segmentAssumedDays > 0 {
                quality = .estimated
            } else if segmentLoads.count < coverageWindow {
                quality = .provisional
            } else {
                quality = .observed
            }

            points.append(Point(
                day: item.day,
                observedLoad: item.observed,
                effectiveLoad: effective,
                source: source,
                atl: atl,
                ctl: ctl,
                tsb: tsb,
                currentRampWindowLoad: ramp.current,
                previousRampWindowLoad: ramp.previous,
                rampChange: ramp.change,
                rampChangeFraction: ramp.fraction,
                rampDirection: ramp.direction,
                observedCoverage: coverage,
                assumedRestDaysInWindow: assumedCount,
                assumedRestDaysInModelHistory: segmentAssumedDays,
                consecutiveHistoryDays: segmentLoads.count,
                quality: quality,
                interpretation: interpretation(quality: quality, tsb: tsb,
                                               consecutiveDays: segmentLoads.count,
                                               minimumDays: configuration.minimumHistoryDays,
                                               assumedCount: segmentAssumedDays)
            ))
        }

        return Series(observedDays: aggregation.days, points: points,
                      rejectedEntryCount: aggregation.rejectedEntryCount, status: .complete)
    }

    private static func rampContext(effectiveLoads: [Double?], windowDays: Int,
                                    stableFraction: Double) ->
        (current: Double?, previous: Double?, change: Double?, fraction: Double?, direction: RampDirection) {
        // Division-first avoids `windowDays * 2` overflow even if this private helper is ever reused
        // independently of Configuration's public bounds.
        guard windowDays > 0, windowDays <= effectiveLoads.count / 2 else {
            return (nil, nil, nil, nil, .unavailable)
        }
        let currentStart = effectiveLoads.count - windowDays
        let previousStart = currentStart - windowDays
        let previousSlice = effectiveLoads[previousStart..<currentStart]
        let currentSlice = effectiveLoads[currentStart...]
        guard previousSlice.allSatisfy({ $0 != nil }), currentSlice.allSatisfy({ $0 != nil }) else {
            return (nil, nil, nil, nil, .unavailable)
        }
        guard let previous = finiteSum(previousSlice.compactMap { $0 }),
              let current = finiteSum(currentSlice.compactMap { $0 }) else {
            return (nil, nil, nil, nil, .unavailable)
        }
        let change = current - previous
        guard change.isFinite else { return (nil, nil, nil, nil, .unavailable) }
        guard previous > 0 else {
            let direction: RampDirection = current > 0 ? .building : .steady
            return (current, previous, change, nil, direction)
        }
        let fraction = change / previous
        guard fraction.isFinite else {
            return (current, previous, change, nil, .unavailable)
        }
        let direction: RampDirection
        if fraction > stableFraction {
            direction = .building
        } else if fraction < -stableFraction {
            direction = .easing
        } else {
            direction = .steady
        }
        return (current, previous, change, fraction, direction)
    }

    private static func interpretation(quality: Quality, tsb: Double?, consecutiveDays: Int,
                                       minimumDays: Int, assumedCount: Int) -> String {
        guard let tsb, tsb.isFinite else {
            return "Load comparison unavailable: \(consecutiveDays) of \(minimumDays) consecutive calendar days recorded."
        }
        if quality == .estimated {
            return "Estimate includes \(assumedCount) unobserved day\(assumedCount == 1 ? "" : "s") treated as rest; it describes recorded load only."
        }
        let comparison: String
        if tsb < -0.000_001 {
            comparison = "Recent recorded load is above the longer-term recorded-load baseline."
        } else if tsb > 0.000_001 {
            comparison = "Recent recorded load is below the longer-term recorded-load baseline."
        } else {
            comparison = "Recent and longer-term recorded load are similar."
        }
        if quality == .provisional {
            return "Provisional: \(comparison) This is load context, not a fitness or fatigue measurement."
        }
        return "\(comparison) This is load context, not a fitness or fatigue measurement."
    }

    /// Incremental mean for nonnegative finite loads. Unlike `sum / count`, this cannot overflow merely
    /// because several individually representable daily values are near `Double.greatestFiniteMagnitude`.
    private static func finiteMean(_ values: [Double]) -> Double? {
        guard !values.isEmpty else { return nil }
        var mean = 0.0
        for (index, value) in values.enumerated() {
            guard value.isFinite, value >= 0 else { return nil }
            let next = mean + (value - mean) / Double(index + 1)
            guard next.isFinite else { return nil }
            mean = next
        }
        return mean
    }

    /// A ramp total is itself part of the public result, so a mathematically unrepresentable sum is
    /// unavailable rather than Infinity. Inputs are already nonnegative, making overflow monotonic.
    private static func finiteSum(_ values: [Double]) -> Double? {
        var total = 0.0
        for value in values {
            guard value.isFinite, value >= 0 else { return nil }
            let next = total + value
            guard next.isFinite else { return nil }
            total = next
        }
        return total
    }

    private static var utcCalendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "en_US_POSIX")
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }

    private static func parseDay(_ value: String) -> Date? {
        let parts = value.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3, parts[0].count == 4, parts[1].count == 2, parts[2].count == 2,
              let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]) else { return nil }
        let calendar = utcCalendar
        guard let date = calendar.date(from: DateComponents(year: year, month: month, day: day)),
              dayKey(date) == value else { return nil }
        return date
    }

    private static func dayKey(_ date: Date) -> String {
        let parts = utcCalendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year!, parts.month!, parts.day!)
    }
}
